File: src/main/java/org/zwobble/mammoth/internal/util/PassThroughException.java
package org.zwobble.mammoth.internal.util;

import java.io.IOException;

public class PassThroughException extends RuntimeException {
    public interface SupplierWithException<T, E extends Throwable> {
        public T get() throws E;
    }

    public static <T> T wrap(SupplierWithException<T, IOException> supplier) {
        try {
            return supplier.get();
        } catch (IOException exception) {
            throw new PassThroughException(exception);
        }
    }

    public static <T> T unwrap(SupplierWithException<T, IOException> supplier) throws IOException {
        try {
            return supplier.get();
        } catch (PassThroughException exception) {
            throw exception.exception;
        }
    }

    private final IOException exception;

    public PassThroughException(IOException exception) {
        super(exception);
        this.exception = exception;
    }
}

File: dotnet/Mammoth/Couscous/mammoth/internal/util/PassThroughException.cs
using Mammoth.Couscous.java.lang;
using Mammoth.Couscous.java.io;

namespace Mammoth.Couscous.org.zwobble.mammoth.@internal.util {
	internal class PassThroughException : RuntimeException {
		internal static T wrap<T>(SupplierWithException<T, IOException> supplier) {
			try {
				return supplier.get();
			} catch (IOException exception) {
				throw new PassThroughException(exception);
			}
		}

		internal static T unwrap<T>(SupplierWithException<T, IOException> supplier) {
			try {
				return supplier.get();
			} catch (PassThroughException exception) {
				throw exception._exception;
			}
		}
    
		private readonly IOException _exception;
		
		internal PassThroughException(IOException exception) {
			_exception = exception;
		}
	}
}

